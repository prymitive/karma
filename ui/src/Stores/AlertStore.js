import { observable, action, computed, toJS } from "mobx";

import throttle from "lodash.throttle";

import qs from "qs";

import { FetchGet } from "Common/Fetch";

const QueryStringEncodeOptions = {
  encodeValuesOnly: true, // don't encode q[]
  indices: false, // go-gin doesn't support parsing q[0]=foo&q[1]=bar
};

function FormatAlertsQ(filters) {
  return qs.stringify({ q: filters }, QueryStringEncodeOptions);
}

// generate URL for the UI with a set of filters
function FormatAPIFilterQuery(filters) {
  return qs.stringify(
    Object.assign(DecodeLocationSearch(window.location.search).params, {
      q: filters,
    }),
    QueryStringEncodeOptions
  );
}

// format URI for react UI -> Go backend requests
function FormatBackendURI(path) {
  const uri = process.env.REACT_APP_BACKEND_URI || ".";
  return `${uri}/${path}`;
}

// takes the '?foo=bar&foo=baz' part of http://example.com?foo=bar&foo=baz
// and decodes it into a dict with some extra metadata
function DecodeLocationSearch(searchString) {
  let defaultsUsed = true;
  let params = { q: [] };

  if (searchString !== "") {
    const parsed = qs.parse(searchString.split("?")[1]);
    params = Object.assign(params, parsed);

    if (parsed.q !== undefined) {
      defaultsUsed = false;
      if (parsed.q === "") {
        params.q = [];
      } else if (Array.isArray(parsed.q)) {
        // first filter out duplicates
        // then filter out empty strings, so 'q=' doesn't end up [""] but rather []
        params.q = parsed.q
          .filter((v, i) => parsed.q.indexOf(v) === i)
          .filter((v) => v !== "");
      } else {
        params.q = [parsed.q];
      }
    }
  }

  return { params: params, defaultsUsed: defaultsUsed };
}

function UpdateLocationSearch(newParams) {
  const baseURLWithoutSearch = window.location.href.split("?")[0];
  const newSearch = FormatAPIFilterQuery(newParams.q);
  window.history.pushState(
    null,
    null,
    `${baseURLWithoutSearch}?${newSearch || "q="}`
  );
}

const AlertStoreStatuses = Object.freeze({
  Idle: Symbol("idle"),
  Fetching: Symbol("fetching"),
  Processing: Symbol("processing"),
  Failure: Symbol("failure"),
});

function NewUnappliedFilter(raw) {
  return {
    applied: false,
    isValid: true,
    raw: raw,
    hits: 0,
    name: "",
    matcher: "",
    value: "",
  };
}

class AlertStore {
  filters = observable(
    {
      values: [],
      addFilter(raw) {
        if (this.values.filter((f) => f.raw === raw).length === 0) {
          this.values.push(NewUnappliedFilter(raw));
          UpdateLocationSearch({ q: this.values.map((f) => f.raw) });
        }
      },
      removeFilter(raw) {
        if (this.values.filter((f) => f.raw === raw).length > 0) {
          this.values = this.values.filter((f) => f.raw !== raw);
          UpdateLocationSearch({ q: this.values.map((f) => f.raw) });
        }
      },
      replaceFilter(oldRaw, newRaw) {
        const index = this.values.findIndex((e) => e.raw === oldRaw);
        if (index >= 0) {
          // first check if we would create a duplicated filter
          if (this.values.findIndex((e) => e.raw === newRaw) >= 0) {
            // we already have newRaw, simply drop oldRaw
            this.removeFilter(oldRaw);
          } else {
            // no dups, continue with a swap
            this.values[index] = NewUnappliedFilter(newRaw);
            UpdateLocationSearch({ q: this.values.map((f) => f.raw) });
          }
        }
      },
      setFilters(raws) {
        this.values = raws.map((raw) => NewUnappliedFilter(raw));
        UpdateLocationSearch({ q: this.values.map((f) => f.raw) });
      },
      setWithoutLocation(raws) {
        const filtersByRaw = this.values.reduce(function (map, obj) {
          map[toJS(obj.raw)] = toJS(obj);
          return map;
        }, {});
        this.values = raws.map((raw) =>
          filtersByRaw[raw] ? filtersByRaw[raw] : NewUnappliedFilter(raw)
        );
      },
      applyAllFilters() {
        for (let i = 0; i < this.values.length; i++) {
          this.values[i].applied = true;
        }
      },
    },
    {
      addFilter: action.bound,
      removeFilter: action.bound,
      replaceFilter: action.bound,
      setFilters: action.bound,
      setWithoutLocation: action.bound,
      applyAllFilters: action.bound,
    },
    { name: "API Filters" }
  );

  data = observable(
    {
      colors: {},
      counters: [],
      grids: [],
      silences: {},
      upstreams: { instances: [], clusters: {} },
      receivers: [],
      getAlertmanagerByName(name) {
        return this.upstreams.instances.find((am) => am.name === name);
      },
      isReadOnlyAlertmanager(name) {
        return this.readOnlyAlertmanagers.map((am) => am.name).includes(name);
      },
      getClusterAlertmanagersWithoutReadOnly(clusterID) {
        return this.clustersWithoutReadOnly[clusterID] || [];
      },
      get readOnlyAlertmanagers() {
        return this.upstreams.instances.filter((am) => am.readonly === true);
      },
      get readWriteAlertmanagers() {
        return this.upstreams.instances
          .filter((am) => am.readonly === false)
          .map((am) =>
            Object.assign({}, am, {
              clusterMembers: am.clusterMembers.filter(
                (m) => this.isReadOnlyAlertmanager(m) === false
              ),
            })
          );
      },
      get clustersWithoutReadOnly() {
        const clusters = {};
        for (const clusterID of Object.keys(this.upstreams.clusters)) {
          const members = this.upstreams.clusters[clusterID].filter(
            (member) => this.isReadOnlyAlertmanager(member) === false
          );
          if (members.length > 0) {
            clusters[clusterID] = members;
          }
        }
        return clusters;
      },
      getColorData(name, value) {
        if (this.colors[name] !== undefined) {
          return this.colors[name][value];
        }
      },
    },
    {
      readOnlyAlertmanagers: computed,
      readWriteAlertmanagers: computed,
      clustersWithoutReadOnly: computed,
    },
    { name: "API Response data" }
  );

  info = observable(
    {
      authentication: {
        enabled: false,
        username: "",
      },
      totalAlerts: 0,
      version: "unknown",
      upgradeNeeded: false,
      isRetrying: false,
      reloadNeeded: false,
      setIsRetrying() {
        this.isRetrying = true;
      },
      clearIsRetrying() {
        this.isRetrying = false;
      },
      setReloadNeeded() {
        this.reloadNeeded = true;
      },
    },
    {
      setIsRetrying: action.bound,
      clearIsRetrying: action.bound,
      setReloadNeeded: action,
    },
    { name: "API response info" }
  );

  settings = observable(
    {
      values: {
        staticColorLabels: [],
        annotationsDefaultHidden: false,
        annotationsHidden: [],
        annotationsVisible: [],
        sorting: {
          grid: {
            order: "startsAt",
            reverse: false,
            label: "alertname",
          },
          valueMapping: {},
        },
        silenceForm: {
          strip: {
            labels: [],
          },
        },
        alertAcknowledgement: {
          enabled: false,
          durationSeconds: 900,
          author: "karma / author missing",
          commentPrefix: "",
        },
      },
    },
    {},
    {
      name: "Global settings",
    }
  );

  status = observable(
    {
      value: AlertStoreStatuses.Idle,
      lastUpdateAt: 0,
      error: null,
      paused: false,
      setIdle() {
        this.value = AlertStoreStatuses.Idle;
        this.error = null;
        this.lastUpdateAt = new Date();
      },
      setFetching() {
        this.value = AlertStoreStatuses.Fetching;
      },
      setProcessing() {
        this.value = AlertStoreStatuses.Processing;
        this.error = null;
      },
      setFailure(err) {
        this.value = AlertStoreStatuses.Failure;
        this.error = err;
        this.lastUpdateAt = new Date();
      },
      pause() {
        this.paused = true;
      },
      resume() {
        this.paused = false;
      },
    },
    {
      setIdle: action,
      setFetching: action,
      setProcessing: action,
      setFailure: action,
      pause: action.bound,
      resume: action.bound,
    },
    { name: "Store status" }
  );

  constructor(initialFilters) {
    if (initialFilters !== null) this.filters.setFilters(initialFilters);
  }

  fetch = action(
    (gridLabel, gridSortReverse, sortOrder, sortLabel, sortReverse) => {
      this.status.setFetching();

      const args = [
        `gridLabel=${gridLabel}`,
        `gridSortReverse=${gridSortReverse ? "1" : "0"}`,
        `sortOrder=${sortOrder}`,
        `sortLabel=${sortLabel}`,
        `sortReverse=${sortReverse}`,
      ];

      const alertsURI =
        FormatBackendURI(`alerts.json?&${args.join("&")}&`) +
        FormatAPIFilterQuery(this.filters.values.map((f) => f.raw));

      return FetchGet(alertsURI, {}, this.info.setIsRetrying)
        .then((result) => {
          // we're sending requests with mode=cors so the response should also be type=cors
          // after a few failures in the retry loop we will switch to no-cors
          // if that request comes back as type=opaque then we might be getting
          // redirected by an auth proxy
          if (result.type === "opaque") {
            this.info.setReloadNeeded();
          }
          this.info.clearIsRetrying();
          this.status.setProcessing();
          return result.json();
        })
        .then((result) => {
          return this.parseAPIResponse(result);
        })
        .catch((err) => {
          console.trace(err);
          return this.handleFetchError(
            `Can't connect to the API, last error was "${err.message}"`
          );
        });
    }
  );

  fetchWithThrottle = throttle(this.fetch, 300);

  parseAPIResponse = action((result) => {
    if (result.error) {
      this.handleFetchError(result.error);
      return;
    }

    const queryFilters = [
      ...new Set(
        this.filters.values
          .map((f) => f.raw)
          .slice()
          .sort()
      ),
    ];
    const responseFilters = [
      ...new Set(result.filters.map((m) => m.text).sort()),
    ];
    if (JSON.stringify(queryFilters) !== JSON.stringify(responseFilters)) {
      console.info(
        `Got response with filters '${responseFilters}' while expecting results for '${queryFilters}', ignoring`
      );
      return;
    }

    for (const filter of result.filters) {
      const storedIndex = this.filters.values.findIndex(
        (f) => f.raw === filter.text
      );
      this.filters.values[storedIndex] = Object.assign(
        this.filters.values[storedIndex],
        {
          applied: true,
          isValid: filter.isValid,
          hits: filter.hits,
          name: filter.name,
          matcher: filter.matcher,
          value: filter.value,
        }
      );
    }

    let updates = {};
    for (const key of [
      "colors",
      "counters",
      "grids",
      "silences",
      "upstreams",
      "receivers",
    ]) {
      updates[key] = result[key];
    }
    this.data = Object.assign(this.data, updates);

    // before storing new version check if we need to reload
    if (
      this.info.version !== "unknown" &&
      this.info.version !== result.version
    ) {
      this.info.upgradeNeeded = true;
    }
    // update extra root level keys that are stored under 'info'
    for (const key of ["totalAlerts", "version", "authentication"]) {
      if (this.info[key] !== result[key]) {
        this.info[key] = result[key];
      }
    }

    // settings exported via API
    this.settings.values = result.settings;

    this.status.setIdle();
  });

  handleFetchError = action((err) => {
    this.status.setFailure(err);

    // reset alert counter since we won't be rendering any alerts
    this.info.totalAlerts = 0;

    // all unapplied filters should be marked applied to reset progress indicator
    this.filters.applyAllFilters();

    return { error: err };
  });
}

export {
  AlertStore,
  AlertStoreStatuses,
  FormatBackendURI,
  FormatAPIFilterQuery,
  FormatAlertsQ,
  DecodeLocationSearch,
  UpdateLocationSearch,
  NewUnappliedFilter,
};
