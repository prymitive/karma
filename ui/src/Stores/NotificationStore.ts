import { action, computed, observable, makeObservable } from "mobx";

export interface NotificationT {
  id: string;
  type: "warning" | "error";
  title: string;
  message: string;
  timestamp: Date;
  dismissed: boolean;
  autoDismissed: boolean;
  source: "alertmanager" | "system";
  sourceId?: string; // For alertmanager notifications, this would be the upstream name
  occurrenceCount: number; // Track how many times this notification has occurred
}

class NotificationStore {
  notifications: NotificationT[] = [];

  constructor() {
    makeObservable(this, {
      notifications: observable,
      addNotification: action,
      dismissNotification: action,
      dismissAllBySource: action,
      clearDismissed: action,
      activeNotifications: computed,
      dismissedNotifications: computed,
      warningCount: computed,
      errorCount: computed,
      totalActiveCount: computed,
    });
  }

  addNotification = (
    notification: Omit<
      NotificationT,
      "id" | "timestamp" | "dismissed" | "autoDismissed" | "occurrenceCount"
    >,
  ) => {
    // Check if we already have an active notification for this source
    const existingActiveIndex = this.notifications.findIndex(
      (n) =>
        n.sourceId === notification.sourceId &&
        n.source === notification.source &&
        !n.dismissed,
    );

    // Check if we have a previously dismissed notification for this source
    const existingDismissedIndex = this.notifications.findIndex(
      (n) =>
        n.sourceId === notification.sourceId &&
        n.source === notification.source &&
        n.dismissed,
    );

    if (existingActiveIndex >= 0) {
      // Update existing active notification
      const newNotification: NotificationT = {
        ...notification,
        id: this.notifications[existingActiveIndex].id,
        timestamp: new Date(),
        dismissed: false,
        autoDismissed: false,
        occurrenceCount:
          this.notifications[existingActiveIndex].occurrenceCount + 1,
      };
      this.notifications[existingActiveIndex] = newNotification;
    } else if (existingDismissedIndex >= 0) {
      // Instead of creating a new notification, increment the count on the dismissed one
      const dismissedNotification = this.notifications[existingDismissedIndex];
      dismissedNotification.occurrenceCount += 1;
      dismissedNotification.timestamp = new Date(); // Update timestamp to reflect latest occurrence
      // Note: We keep it dismissed but show the updated count in history
    } else {
      // Add new notification
      const newNotification: NotificationT = {
        ...notification,
        id: `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
        timestamp: new Date(),
        dismissed: false,
        autoDismissed: false,
        occurrenceCount: 1,
      };
      this.notifications.push(newNotification);
    }

    // Keep only last 50 notifications to prevent memory issues
    if (this.notifications.length > 50) {
      this.notifications = this.notifications.slice(-50);
    }
  };

  dismissNotification = (id: string, isAutoDismiss = false) => {
    const notification = this.notifications.find((n) => n.id === id);
    if (notification) {
      notification.dismissed = true;
      notification.autoDismissed = isAutoDismiss;
    }
  };

  dismissAllBySource = (
    source: string,
    sourceId?: string,
    isAutoDismiss = false,
  ) => {
    this.notifications.forEach((notification) => {
      if (
        notification.source === source &&
        (!sourceId || notification.sourceId === sourceId) &&
        !notification.dismissed
      ) {
        notification.dismissed = true;
        notification.autoDismissed = isAutoDismiss;
      }
    });
  };

  clearDismissed = () => {
    this.notifications = this.notifications.filter((n) => !n.dismissed);
  };

  get activeNotifications(): NotificationT[] {
    return this.notifications.filter((n) => !n.dismissed);
  }

  get dismissedNotifications(): NotificationT[] {
    return this.notifications.filter((n) => n.dismissed).slice(-20); // Keep last 20 dismissed
  }

  get warningCount(): number {
    return this.activeNotifications.filter((n) => n.type === "warning").length;
  }

  get errorCount(): number {
    return this.activeNotifications.filter((n) => n.type === "error").length;
  }

  get totalActiveCount(): number {
    return this.activeNotifications.length;
  }
}

export { NotificationStore };
export const notificationStore = new NotificationStore();
