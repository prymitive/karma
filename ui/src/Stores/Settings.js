import { action } from "mobx";
import { localStored } from "mobx-stored";

class SavedFilters {
  config = localStored(
    "savedFilters",
    {
      filters: [],
      present: false
    },
    {
      delay: 100
    }
  );

  save = action(newFilters => {
    this.config.filters = newFilters;
    this.config.present = true;
  });

  clear = action(() => {
    this.config.filters = [];
    this.config.present = false;
  });
}

class FetchConfig {
  constructor(refresh) {
    this.config = localStored(
      "fetchConfig",
      { interval: refresh },
      { delay: 100 }
    );

    this.setInterval = action(newInterval => {
      this.config.interval = newInterval;
    });
  }
}

class AlertGroupConfig {
  options = Object.freeze({
    expanded: { label: "Always expanded", value: "expanded" },
    collapsedOnMobile: {
      label: "Collapse on mobile",
      value: "collapsedOnMobile"
    },
    collapsed: { label: "Always collapsed", value: "collapsed" }
  });

  constructor(renderCount, collapseState, colorTitleBar) {
    this.config = localStored(
      "alertGroupConfig",
      {
        defaultRenderCount: renderCount,
        defaultCollapseState: collapseState,
        colorTitleBar: colorTitleBar
      },
      { delay: 100 }
    );
  }

  update = action(data => {
    for (const [key, val] of Object.entries(data)) {
      this.config[key] = val;
    }
  });
}

class SilenceFormConfig {
  config = localStored("silenceFormConfig", { author: "" }, { delay: 100 });

  saveAuthor = action(newAuthor => {
    this.config.author = newAuthor;
  });
}

class GridConfig {
  options = Object.freeze({
    default: { label: "Use defaults from karma config file", value: "default" },
    disabled: { label: "No sorting", value: "disabled" },
    startsAt: { label: "Sort by alert timestamp", value: "startsAt" },
    label: { label: "Sort by alert label", value: "label" }
  });
  constructor(groupWidth) {
    this.config = localStored(
      "alertGridConfig",
      {
        sortOrder: this.options.default.value,
        sortLabel: null,
        reverseSort: null,
        groupWidth: groupWidth
      },
      { delay: 100 }
    );
  }
}

class FilterBarConfig {
  constructor(autohide) {
    this.config = localStored(
      "filterBarConfig",
      {
        autohide: autohide
      },
      {
        delay: 100
      }
    );
  }
}

class Settings {
  constructor(defaults) {
    let defaultSettings;
    if (defaults === undefined) {
      defaultSettings = {
        Refresh: 30 * 1000 * 1000 * 1000,
        HideFiltersWhenIdle: true,
        ColorTitlebar: false,
        MinimalGroupWidth: 420,
        AlertsPerGroup: 5,
        CollapseGroups: "collapsedOnMobile"
      };
    } else {
      defaultSettings = defaults;
    }

    this.savedFilters = new SavedFilters();
    this.fetchConfig = new FetchConfig(
      defaultSettings.Refresh / 1000 / 1000 / 1000
    );
    this.alertGroupConfig = new AlertGroupConfig(
      defaultSettings.AlertsPerGroup,
      defaultSettings.CollapseGroups,
      defaultSettings.ColorTitlebar
    );
    this.gridConfig = new GridConfig(defaultSettings.MinimalGroupWidth);
    this.silenceFormConfig = new SilenceFormConfig();
    this.filterBarConfig = new FilterBarConfig(
      defaultSettings.HideFiltersWhenIdle
    );
  }
}

export { Settings };
