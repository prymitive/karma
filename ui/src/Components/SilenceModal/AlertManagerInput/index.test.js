import React from "react";

import { shallow, mount } from "enzyme";

import toDiffableHtml from "diffable-html";

import { AlertStore } from "Stores/AlertStore";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { AlertManagerInput } from ".";

let alertStore;
let silenceFormStore;

beforeEach(() => {
  alertStore = new AlertStore([]);
  alertStore.data.upstreams.clusters = {
    ha: ["am1", "am2"],
    am3: ["am3"]
  };
  alertStore.data.upstreams.instances = [
    {
      name: "am1",
      uri: "http://am1.example.com",
      publicURI: "http://am1.example.com",
      error: "",
      version: "0.15.0",
      cluster: "ha",
      clusterMembers: ["am1", "am2"]
    },
    {
      name: "am2",
      uri: "http://am2.example.com",
      publicURI: "http://am2.example.com",
      error: "",
      version: "0.15.0",
      cluster: "ha",
      clusterMembers: ["am1", "am2"]
    },
    {
      name: "am3",
      uri: "http://am3.example.com",
      publicURI: "http://am3.example.com",
      error: "",
      version: "0.15.0",
      cluster: "am3",
      clusterMembers: ["am3"]
    }
  ];
  silenceFormStore = new SilenceFormStore();
});

const ShallowAlertManagerInput = () => {
  return shallow(
    <AlertManagerInput
      alertStore={alertStore}
      silenceFormStore={silenceFormStore}
    />
  );
};

const MountedAlertManagerInput = () => {
  return mount(
    <AlertManagerInput
      alertStore={alertStore}
      silenceFormStore={silenceFormStore}
    />
  );
};

const ValidateSuggestions = () => {
  const tree = MountedAlertManagerInput();
  // clear all selected instances, they are selected by default
  const clear = tree.find("ClearIndicator");
  // https://github.com/JedWatson/react-select/blob/c22d296d50917e210836fb011ae3e565895e6440/src/__tests__/Select.test.js#L1873
  clear.simulate("mousedown", { button: 0 });
  // click on the react-select component doesn't seem to trigger options
  // rendering in tests, so change the input instead
  tree.find("input").simulate("change", { target: { value: "am" } });
  return tree;
};

describe("<AlertManagerInput />", () => {
  it("matches snapshot", () => {
    const tree = ShallowAlertManagerInput();
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("doesn't render ValidationError after passed validation", () => {
    const tree = ShallowAlertManagerInput();
    silenceFormStore.data.wasValidated = true;
    expect(tree.html()).not.toMatch(/fa-exclamation-circle/);
    expect(tree.html()).not.toMatch(/Required/);
  });

  it("renders ValidationError after failed validation", () => {
    const tree = ShallowAlertManagerInput();
    silenceFormStore.data.alertmanagers = [];
    silenceFormStore.data.wasValidated = true;
    expect(tree.html()).toMatch(/fa-exclamation-circle/);
    expect(tree.html()).toMatch(/Required/);
  });

  it("all available Alertmanager instances are selected by default", () => {
    ShallowAlertManagerInput();
    expect(silenceFormStore.data.alertmanagers).toHaveLength(2);
    expect(silenceFormStore.data.alertmanagers).toContainEqual({
      label: "am1 | am2",
      value: ["am1", "am2"]
    });
    expect(silenceFormStore.data.alertmanagers).toContainEqual({
      label: "am3",
      value: ["am3"]
    });
  });

  it("doesn't override last selected Alertmanager instances on mount", () => {
    silenceFormStore.data.alertmanagers = [{ label: "am3", value: ["am3"] }];
    ShallowAlertManagerInput();
    expect(silenceFormStore.data.alertmanagers).toHaveLength(1);
    expect(silenceFormStore.data.alertmanagers).toContainEqual({
      label: "am3",
      value: ["am3"]
    });
  });

  it("renders all 3 suggestions", () => {
    const tree = ValidateSuggestions();
    const options = tree.find("[role='option']");
    expect(options).toHaveLength(2);
    expect(options.at(0).text()).toBe("am1 | am2");
    expect(options.at(1).text()).toBe("am3");
  });

  it("clicking on options appends them to silenceFormStore.data.alertmanagers", () => {
    silenceFormStore.data.alertmanagers = [];
    const tree = ValidateSuggestions();
    const options = tree.find("[role='option']");
    options.at(0).simulate("click");
    options.at(1).simulate("click");
    expect(silenceFormStore.data.alertmanagers).toHaveLength(2);
    expect(silenceFormStore.data.alertmanagers).toContainEqual({
      label: "am1 | am2",
      value: ["am1", "am2"]
    });
    expect(silenceFormStore.data.alertmanagers).toContainEqual({
      label: "am3",
      value: ["am3"]
    });
  });

  it("silenceFormStore.data.alertmanagers gets updated from alertStore.data.upstreams.instances on mismatch", () => {
    const tree = ShallowAlertManagerInput();
    alertStore.data.upstreams.clusters = {
      amNew: ["amNew"]
    };
    // force update since this is where the mismatch check lives
    tree.instance().componentDidUpdate();
    expect(silenceFormStore.data.alertmanagers).toContainEqual({
      label: "amNew",
      value: ["amNew"]
    });
  });

  it("is enabled when silenceFormStore.data.silenceID is null", () => {
    silenceFormStore.data.silenceID = null;
    const tree = MountedAlertManagerInput();
    const select = tree.find("StateManager");
    expect(select.props().isDisabled).toBeFalsy();
  });

  it("is disabled when silenceFormStore.data.silenceID is not null", () => {
    silenceFormStore.data.silenceID = "1234";
    const tree = MountedAlertManagerInput();
    const select = tree.find("StateManager");
    expect(select.props().isDisabled).toBe(true);
  });
});
