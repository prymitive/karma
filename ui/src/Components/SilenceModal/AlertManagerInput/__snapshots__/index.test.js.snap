// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlertManagerInput /> matches snapshot 1`] = `
"
<div class=\\"css-10nd86i\\">
  <div class=\\"css-7jxtyj\\">
    <div class=\\"css-10war8y\\">
      <div class=\\"css-1y5uxcf\\">
        <div class=\\"css-yagan3\\">
          am1
        </div>
        <div class=\\"css-n82uvk\\">
          <svg height=\\"14\\"
               width=\\"14\\"
               viewbox=\\"0 0 20 20\\"
               aria-hidden=\\"true\\"
               focusable=\\"false\\"
               class=\\"css-19bqh2r\\"
          >
            <path d=\\"M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z\\">
            </path>
          </svg>
        </div>
      </div>
      <div class=\\"css-1y5uxcf\\">
        <div class=\\"css-yagan3\\">
          am2
        </div>
        <div class=\\"css-n82uvk\\">
          <svg height=\\"14\\"
               width=\\"14\\"
               viewbox=\\"0 0 20 20\\"
               aria-hidden=\\"true\\"
               focusable=\\"false\\"
               class=\\"css-19bqh2r\\"
          >
            <path d=\\"M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z\\">
            </path>
          </svg>
        </div>
      </div>
      <div class=\\"css-1y5uxcf\\">
        <div class=\\"css-yagan3\\">
          am3
        </div>
        <div class=\\"css-n82uvk\\">
          <svg height=\\"14\\"
               width=\\"14\\"
               viewbox=\\"0 0 20 20\\"
               aria-hidden=\\"true\\"
               focusable=\\"false\\"
               class=\\"css-19bqh2r\\"
          >
            <path d=\\"M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z\\">
            </path>
          </svg>
        </div>
      </div>
      <div class=\\"css-1g6gooi\\">
        <div class
             style=\\"display:inline-block\\"
        >
          <input type=\\"text\\"
                 autocapitalize=\\"none\\"
                 autocomplete=\\"off\\"
                 autocorrect=\\"off\\"
                 id=\\"react-select-silence-input-alertmanagers-input\\"
                 spellcheck=\\"false\\"
                 tabindex=\\"0\\"
                 value
                 aria-autocomplete=\\"list\\"
                 style=\\"box-sizing:content-box;width:1px;background:0;border:0;font-size:inherit;opacity:1;outline:0;padding:0;color:inherit\\"
          >
          <div style=\\"position:absolute;top:0;left:0;visibility:hidden;height:0;overflow:scroll;white-space:pre\\">
          </div>
        </div>
      </div>
    </div>
    <div class=\\"css-mik995\\">
      <div aria-hidden=\\"true\\"
           class=\\"css-1ep9fjw\\"
      >
        <svg height=\\"20\\"
             width=\\"20\\"
             viewbox=\\"0 0 20 20\\"
             aria-hidden=\\"true\\"
             focusable=\\"false\\"
             class=\\"css-19bqh2r\\"
        >
          <path d=\\"M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z\\">
          </path>
        </svg>
      </div>
      <span class=\\"css-d8oujb\\">
      </span>
      <div aria-hidden=\\"true\\"
           class=\\"css-1ep9fjw\\"
      >
        <svg height=\\"20\\"
             width=\\"20\\"
             viewbox=\\"0 0 20 20\\"
             aria-hidden=\\"true\\"
             focusable=\\"false\\"
             class=\\"css-19bqh2r\\"
        >
          <path d=\\"M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z\\">
          </path>
        </svg>
      </div>
    </div>
  </div>
</div>
"
`;
